File: packages/dirigent-server/src/dirigent_server/routes/system.py
"""What this instance is, and what each connection said when it was last checked."""

import sqlalchemy as sa
from fastapi import APIRouter

from dirigent_client.schemas import ConnectionHealth, SystemInfo
from dirigent_core import __version__
from dirigent_core.models import Connection, Worker, utcnow
from dirigent_core.registry import STALE_AFTER
from dirigent_server.dependencies import ServicesDep, SessionDep, SettingsDep
from dirigent_server.security import PrincipalDep
from dirigent_server.transactions import Transactional

router = APIRouter(route_class=Transactional, tags=["system"])


@router.get(
    "/system/info",
    operation_id="getSystemInfo",
    summary="Describe the instance and repeat every connection's last check",
    response_model=SystemInfo,
)
async def system_info(
    session: SessionDep,
    services: ServicesDep,
    settings: SettingsDep,
    principal: PrincipalDep,
) -> SystemInfo:
    """Describe the instance from what it holds, without opening any connection.

    A connection's health is what its last check wrote on the row: this read is made on every
    page load, and must not run anyone's connect timeout.
    """
    catalog = services.host.catalog()
    rows = list((await session.execute(sa.select(Connection).order_by(Connection.code))).scalars())
    live = await session.execute(
        sa.select(sa.func.count()).select_from(Worker).where(Worker.last_seen_at >= utcnow() - STALE_AFTER)
    )
    return SystemInfo(
        version=__version__,
        environment=settings.environment,
        database="sqlite" if settings.is_sqlite else "postgresql",
        checked_at=utcnow(),
        plugins=catalog.plugins,
        blocks=len(catalog.blocks),
        storage_schemes=services.storage.schemes,
        notifiers=[entry.id for entry in catalog.notifiers],
        connection_kinds=[entry.id for entry in catalog.connection_kinds],
        workers_live=int(live.scalar_one()),
        unsafe_blocks_enabled=list(settings.enabled_unsafe_blocks),
        secrets_configured=services.secrets.available,
        connections=[
            ConnectionHealth(
                code=row.code,
                name=row.name,
                kind=row.kind,
                last_check_at=row.last_check_at,
                last_check_healthy=row.last_check_healthy,
                last_check_detail=row.last_check_detail,
            )
            for row in rows
        ],
    )

File: packages/dirigent-server/frontend/src/lib/system.ts
/**
 * What this instance has installed.
 *
 * ONE READ, AND IT IS NOT THE PROBE. `GET /system/info` is the versioned API's and says what is
 * installed -- version, environment, database, blocks, storage schemes, plugins, and what every
 * configured connection said when it was last checked. It opens no connection itself: a read
 * made on every page load must not run anyone's connect timeout. Whether the process is fit
 * to serve is a different question
 * with a different answer, asked at the root by `lib/server-status`, and there is exactly one
 * reader of it: the corner dot's store, which the settings dialog reads rather than asking a
 * second time.
 */

import { apiJson } from '@/lib/api'

/** What one configured connection said the last time it was checked. `ConnectionHealth`. */
export interface ConnectionHealth {
    code: string
    name: string | null
    kind: string
    last_check_at: string | null
    last_check_healthy: boolean | null
    last_check_detail: string | null
}

/** Everything this instance is. `SystemInfo`. */
export interface SystemInfo {
    name: string
    version: string
    environment: string
    /** The dialect in use, such as `sqlite` or `postgresql`. */
    database: string
    checked_at: string
    plugins: string[]
    blocks: number
    storage_schemes: string[]
    notifiers: string[]
    connection_kinds: string[]
    workers_live: number
    unsafe_blocks_enabled: string[]
    secrets_configured: boolean
    connections: ConnectionHealth[]
}

/** Read what this instance is. */
export function readSystemInfo(): Promise<SystemInfo> {
    return apiJson<SystemInfo>('/system/info')
}
